api/views/canteen.py

```diff
@@ -58,7 +58,7 @@
 class PublishedCanteenSingleView(RetrieveAPIView):
     model = Canteen
     serializer_class = PublicCanteenSerializer
-    queryset = Canteen.objects.filter(publication_status="published")
+    queryset = Canteen.objects.publicly_visible()
```

Comment:
```suggestion
    def get_queryset(self):
        return Canteen.objects.publicly_visible()
```

Reply:
@christophehenry @hfroot  Ça ne marche pas tout à fait comme ça... Les querysets sont lazy, le fait de le définir à ce niveau ne l’exécute pas.

De plus, dans le code `GenericAPIView` de DRF `get_queryset` est déjà défini et regarde déjà dans la propriété `queryset`. Voici le code en question : 

```python
class GenericAPIView(views.APIView):
    ...
    def get_queryset(self):
        """
        Get the list of items for this view.
        This must be an iterable, and may be a queryset.
        Defaults to using `self.queryset`.

        This method should always be used rather than accessing `self.queryset`
        directly, as `self.queryset` gets evaluated only once, and those results
        are cached for all subsequent requests.

        You may want to override this if you need to provide different
        querysets depending on the incoming request.

        (Eg. return a list of items that is specific to the user)
        """
        assert self.queryset is not None, (
            "'%s' should either include a `queryset` attribute, "
            "or override the `get_queryset()` method."
            % self.__class__.__name__
        )

        queryset = self.queryset
        if isinstance(queryset, QuerySet):
            # Ensure queryset is re-evaluated on each request.
            queryset = queryset.all()
        return queryset
``` 

C'est donc au moment du `queryset.all()` que la requête est vraiment effectuée. 

Comme indiqué d'ailleurs dans le commentaries, la seule raison d'utiliser `get_queryset` est dans le cas où l'on a besoin des informations dans la requête pour définir le queryset (par exemple, retourner des items différents en dépendant du rôle de l'user).

Je préfererais donc de garder `queryset` pour les cas où l'on ne dépend pas des informations contenues dans la requête, comme celui-ci

Reply:
Les queryset eux-mêmes sont lazy. Mais l'évaluation de `Canteen.objects.publicly_visible()` ne l'est pas.